python/openassetio/hostAPI/Manager.py

```diff
@@ -1077,6 +1077,15 @@ def register(self, targetEntityRefs, entitySpecs, context):
         if len(targetEntityRefs) != len(entitySpecs):
             raise IndexError("Parameter lists must be of the same length")
 
+        if entitySpecs:
+            # Check supplied specifcations share a trait set
+            expectedTraits = entitySpecs[0].traitIds()
```

Comment:
[question] is `next(iter(entitySpecs))` preferred here, and manually tracking an index, on the off chance that this isn't a `list` or `tuple`?

Reply:
An intriguing problem. A Pythonic one-liner, but doesn't short-circuit

```python
len(set(frozenset(s.traitIds()) for s in entitySpecs)) == 1
```

Ideally `traitIds()` would return a `frozenset` (needed to be used in a set of sets).

A quick google reveals solutions using `itertools.groupby`, though I suspect some `frozenset` shenanigans would be required for us, [e.g. from StackOverflow](https://stackoverflow.com/a/3844832/535103)

```python
from itertools import groupby

def all_equal(iterable):
    g = groupby(iterable)
    return next(g, True) and not next(g, False)
````

Reply:
Ah but you're trying to make a nice error message, which limits the solution.

Reply:
I think it's reasonable to assume, maybe even require, a sequential container that is indexable, so elements can correspond to elements in `targetEntityRefs`.  The only counterexample that springs to mind is a `dict` whose keys correspond, but that would require different treatment in the manager plugin, complicating matters.